sycl/doc/design/ESIMDDesignNotes.md

```diff
@@ -0,0 +1,277 @@
+the function name is `sycl::accessor::__init`, for ESIMD it is
+`sycl::accessor::__init_esimd`. The difference is caused by the ESIMD limitation
+\- it does not support offset, memory and access range for an accessor object.
+
+There are also couple ESIMD-specific proxy classes used by ESIMD API
+implementation to access internals of ESIMD objects:
+
+- `sycl::ext::intel::esimd::detail::AccessorPrivateProxy`
+  this one is used by ESIMD memory APIs to access internals of `sycl::accessor`
+  objects
+- `sycl::ext::intel::esimd::detail::WrapperElementTypeProxy`
+  used to access internals of SYCL types such as `sycl::detail::half_impl::half`
+
+#### ESIMD API headers
+
+These headers define ESIMD APIs to be used by ESIMD user kernels. For example,
+the basic vector data type `sycl::ext::intel::esimd::simd`,
+`sycl::ext::intel::esimd::gather` memory APIs
+
+Source locations:
+
+- `sycl/ext/intel/esimd`
+- `sycl/ext/intel/experimental/esimd`
+
+### Clang driver
+
+TODO: describe driver modifications.
+
+Source locations:
+
+- `clang/lib/Driver/ToolChains/Clang.cpp`
+- `clang/include/clang/Driver/Options.td`
+
+### Clang front-end
+
+#### Semantic analyzer, integration header generator
+
+- Generation of ESIMD kernel signature and prolog code (which re-assembles
+  objects such as accessors from incoming arguments) is tweaked for ESIMD
+  kernels:
+  - accessors generate fewer kernel arguments
+  - accessor object is assembled in the prolog using `__init_esimd` function
+    instead of `__init`. See function name definition and its usage
+    [here](https://github.com/intel/llvm/blob/eb33bbcfbeab7af1a7f58fb4dc6b53bc47f73dba/clang/lib/Sema/SemaSYCL.cpp#L62).
+- Defines `intel::sycl_explicit_simd` function and variable attribute to markup
+  kernels top-level ESIMD external functions and "private globals".
+- Performs various ESIMD-specific semantic analysis and diagnostics. All kinds
+  of ESIMD error messages can be found in
+  `clang/include/clang/Basic/DiagnosticSemaKinds.td`.
+- Integration header generator inserts additional
+  `static constexpr bool isESIMD()` function into kernel discriptor type, which
+  is used by the host compiler to obtain compile-time details for the kernel via
+  inclusion of the integration header generated by the generator. For ESIMD
+  kernels `isESIMD()` returns true, and this is used in SYCL RT headers to
+  distinguish ESIMD kernels from SYCL kernels, e.g. to customize kernel argument
+  setting for accessors.
+- defines `sycl_esimd_vectorize` attribute propagated to the VC BE, aimed at
+  making it possible to call SIMT function from ESIMD code while vectorizing the
+  former. Used internally to implement certain math functions in SYCL spec
+  conformant manner (proper precision) via vectorizing scalar conformant
+  implementations - see example
+  [usage](https://github.com/intel/llvm/blob/sycl/sycl/include/sycl/ext/intel/esimd/detail/math_intrin.hpp#L199)
```

Comment:
I think we should move this type of documentation the source code and use Doxygen to produce rendered version. It's much more convenient to track the correlation between the code and the documentation. The links and source location pointers are easy to invalidate with code changes.